libraries/standard/http/src/http_client.c

```diff
@@ -245,6 +245,986 @@ static int findHeaderValueParserCallback( http_parser * pHttpParser,
+/* coverity[misra_c_2012_directive_4_6_violation] */
+/* coverity[misra_c_2012_rule_8_13_violation] */
+static int httpParserOnHeadersCompleteCallback( http_parser * pHttpParser )
+{
+    /* The MISRA directive 4.6 violation is for using primitive type int,
+     * instead of a typedef which denotes the signedness and size of the type.
+     * This violation is suppressed because this variable stores the return
+     * value of this callback, which follows the function signature for the
+     * http_data_cb type in http-parser. */
+    /* coverity[misra_c_2012_directive_4_6_violation] */
+    int shouldContinueParse = HTTP_PARSER_CONTINUE_PARSING;
+    HTTPParsingContext_t * pParsingContext = NULL;
+    HTTPResponse_t * pResponse = NULL;
+
+    assert( pHttpParser != NULL );
+    assert( pHttpParser->data != NULL );
+
+    /* The MISRA rule 11.5 violation flags casting a void pointer to another
+     * type. This rule is suppressed here because the http-parser library
+     * requires that private context into callbacks must be set to a void
+     * pointer variable. */
+    /* coverity[misra_c_2012_rule_11_5_violation] */
+    pParsingContext = ( HTTPParsingContext_t * ) ( pHttpParser->data );
+    pResponse = pParsingContext->pResponse;
+
+    assert( pResponse != NULL );
+
+    /* Set the location of what to parse next. */
+    pParsingContext->pBufferCur += HTTP_HEADER_END_INDICATOR_LEN;
+
+    /* If headers existed, then pResponse->pHeaders was set during the first
+     * call to httpParserOnHeaderFieldCallback(). */
+    if( pResponse->pHeaders != NULL )
+    {
+        /* The start of the headers ALWAYS come before the the end of the headers. */
+        assert( ( const char * ) ( pResponse->pHeaders ) < pParsingContext->pBufferCur );
+
+        /* MISRA Rule 10.8 flags the following line for casting from a signed
+         * pointer difference to a size_t. This rule us suppressed because in
+         * in the previous statement it is asserted that the pointer difference
+         * will never be negative. */
+        /* coverity[misra_c_2012_rule_10_8_violation] */
+        pResponse->headersLen = ( size_t ) ( pParsingContext->pBufferCur - ( const char * ) ( pResponse->pHeaders ) );
+    }
+    else
+    {
+        pResponse->headersLen = 0u;
+    }
+
+    /* If the Content-Length header was found, then pHttpParser->content_length
+     * will not be equal to the maximum 64 bit integer. */
+
+    /* MISRA rule 14.3 flags the following line as invaraint. http-parser library
+     * sets pHttpParser->content_length intentionally to the maximum 64 bit
+     * integer to note a non-existent Content-Length header. */
+    /* coverity[misra_c_2012_rule_14_3_violation] */
+    if( pHttpParser->content_length != ( ( uint64_t ) -1 ) )
+    {
+        pParsingContext->pResponse->contentLength = ( size_t ) ( pHttpParser->content_length );
+    }
+    else
+    {
+        pParsingContext->pResponse->contentLength = 0u;
```

Comment:
pParsingContext->pResponse->contentLength can be changed to pResponse->contentLength